platform/FileHandle.h

```diff
@@ -152,8 +156,88 @@ class FileHandle {
      */
     MBED_DEPRECATED_SINCE("mbed-os-5.4", "Replaced by FileHandle::size")
     virtual off_t flen() { return size(); }
+
+    /** Set blocking or non-blocking mode of the file operation like read/write.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  The default is blocking.
+     *
+     *  @param blocking true for blocking mode, false for non-blocking mode.
+     */
+    virtual int set_blocking(bool blocking) {
+        return -1;
+    }
+
+    /** Check for poll event flags
+     * The input parameter can be used or ignored - the could always return all events,
+     * or could check just the events listed in events.
+     * Call is non-blocking - returns instantaneous state of events.
+     * Whenever an event occurs, the derived class must call _poll_change().
+     * @param events bitmask of poll events we're interested in - POLLIN/POLLOUT etc.
+     * @return bitmask of poll events that have occurred.
+     */
+    virtual short poll(short events) const {
+        // Possible default for real files
+        return POLLIN | POLLOUT;
+    }
+    /** Returns true if the FileHandle is writable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, writable() could return
+     *  true when there is ample buffer space available for write() calls.
+     */
+    bool writable() const
+    {
+        return poll(POLLOUT) & POLLOUT;
+    }
+
+    /** Returns true if the FileHandle is readable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, readable() could return
+     *  true when there is something available to read.
+     */
+    bool readable() const
+    {
+        return poll(POLLIN) & POLLIN;
+    }
+
+    /** Register a callback on state change of the file.
+     *
+     *  The specified callback will be called on state changes such as when
+     *  the file can be written to or read from.
+     *
+     *  The callback may be called in an interrupt context and should not
+     *  perform expensive operations.
+     *
+     *  Note! This is not intended as an attach-like asynchronous api, but rather
+     *  as a building block for constructing  such functionality.
+     *
+     *  The exact timing of when the registered function
+     *  is called is not guaranteed and susceptible to change. It should be used
+     *  as a cue to make read/write/poll calls to find the current state.
+     *
+     *  @param func     Function to call on state change
+     */
+    void sigio(Callback<void()> func);
+
+    /** Issue sigio to user - used by mbed::_poll_change */
+    void _send_sigio()
```

Comment:
Well, this review pushed the requirement to store and call a sigio callback onto the implementers - that's more work than simply calling poll_change was. My original design was deliberately simpler than the current version because I do share the concerns about making it excessively fiddly.

Well, actually I care more about scope for getting it subtly wrong or half-working. People can always copy fiddly boilerplate code, it's half-implemented that would be a problem. (They've also got to do their own blocking too - that's a 3rd part of it).

But if proper poll blocking support is a detectable extension, with an automatic fallback to wait_ms() if it's not provided, then it should be okay.

I think we have a general problem that we expect too little of implementers from a system point of view, and we provide them too little. Which means we end up with a bunch of things that sort of work, but not really - Serial spinning hard when it blocks, people creating loads of threads because there's no shared system event queue they can use... And I have a slightly bad feeling that this code will also `wait_ms(1)` forever because we can never agree what colour the wake-up mechanism should be.

Anyway, yes, let's drop the poll_change for now, leaving just the simple sigio.

Looking at CallChain - it's not quite what we want, I think. It does a similar thing but for a different reason. A cond var/wait queue system avoids the dynamic allocation problem, because the wait structure is on the stack of the waiting thread. And is better tuned for the "one-shot" nature of that system - the act of calling automatically and atomically dequeues the waiter, which is what you want for blocking operations.